src/openms/source/ANALYSIS/OPENSWATH/PeakIntegrator.cpp

```diff
@@ -0,0 +1,359 @@
+    return pa;
+  }
+
+  template <typename PeakContainerConstIteratorT>
+  double PeakIntegrator::simpson_(PeakContainerConstIteratorT it_begin, PeakContainerConstIteratorT it_end) const
+  {
+    double integral = 0.0;
+    for (auto it = it_begin + 1; it < it_end - 1; it = it + 2)
+    {
+      const double h = it->getPos() - (it - 1)->getPos();
+      const double k = (it + 1)->getPos() - it->getPos();
+      const double y_h = (it - 1)->getIntensity();
+      const double y_0 = it->getIntensity();
+      const double y_k = (it + 1)->getIntensity();
+      integral += (1.0 / 6.0) * (h + k) * ((2.0 - k / h) * y_h + (pow(h + k, 2) / (h * k)) * y_0 + (2.0 - h / k) * y_k);
+    }
+    return integral;
+  }
+
+  template <typename PeakContainerT>
+  PeakIntegrator::PeakBackground PeakIntegrator::estimateBackground_(
+    const PeakContainerT& p, const double left, const double right,
+    const double peak_apex_pos
+  ) const
+  {
+    const double int_l = p.PosBegin(left)->getIntensity();
+    const double int_r = (p.PosEnd(right) - 1)->getIntensity();
+    const double delta_int = int_r - int_l;
+    const double delta_pos = (p.PosEnd(right) - 1)->getPos() - p.PosBegin(left)->getPos();
+    const double min_int_pos = int_r <= int_l ? (p.PosEnd(right) - 1)->getPos() : p.PosBegin(left)->getPos();
+    const double delta_int_apex = std::fabs(delta_int) * std::fabs(min_int_pos - peak_apex_pos) / delta_pos;
+    double background = 0.0;
+    if (baseline_type_ == BASELINE_TYPE_BASETOBASE)
+    {
+      if (integration_type_ == INTEGRATION_TYPE_TRAPEZOID || integration_type_ == INTEGRATION_TYPE_SIMPSON)
+      {
+        // formula for calculating the background using the trapezoidal rule
+        // background = intensity_min*delta_pos + 0.5*delta_int*delta_pos;
+        background = delta_pos * (std::min(int_r, int_l) + 0.5 * std::fabs(delta_int));
+      }
+      else if (integration_type_ == INTEGRATION_TYPE_INTENSITYSUM)
+      {
+        // calculate the background using the formula
+        // y = mx + b where x = rt or mz, m = slope, b = left intensity
+        // sign of delta_int will determine line direction
+        // background += delta_int / delta_pos * (it->getPos() - left) + int_l;
+        for (auto it = p.PosBegin(left); it != p.PosEnd(right); ++it)
+        {
+          background += it->getPos();
+        }
+        UInt n_points = std::distance(p.PosBegin(left), p.PosEnd(right));
+        background = (background - n_points * p.PosBegin(left)->getPos()) * delta_int / delta_pos + n_points * int_l;
+      }
+    }
+    else if (baseline_type_ == BASELINE_TYPE_VERTICALDIVISION)
+    {
+      if (integration_type_ == INTEGRATION_TYPE_TRAPEZOID || integration_type_ == INTEGRATION_TYPE_SIMPSON)
+      {
+        background = delta_pos * std::min(int_r, int_l);
+      }
+      else if (integration_type_ == INTEGRATION_TYPE_INTENSITYSUM)
+      {
+        for (auto it = p.PosBegin(left); it != p.PosEnd(right); ++it)
```

Comment:
what should this line of code do?

Reply:
That `for` is not useful anymore. It used to count the number of elementes between the iterators (so there was not only `++it`, but also `++count`, which then got removed), before you recommended using `std::distance()`. I'll update the code.